src/main/scala/com/fulcrumgenomics/util/zip/Zipper.scala

```diff
@@ -0,0 +1,840 @@
+    def sizeInBytes(compress: Boolean, capacity: Int = DefaultCapacity): Long = {
+      if (compress) MaxCompressedFixedBlockLength * capacity
+      else DefaultUncompressedFixedBlockLength * capacity
+    }
+  }
+
+  /**
+    * Stores multiple blocks to compress or decompress.  This facilitates more efficient reading and writing
+    * of the data, since more data can be read or written at any one time.  It also allows for multiple blocks to
+    * be compressed or decompressed by a single consumer.
+    *
+    * @param capacity the capacity of fixed-size blocks.
+    * @param compress true if the data will be compressed, false if it is to be decompressed.
+    */
+  private class SequentialBlocks private(capacity: Int, compress: Boolean) {
+    def this(compress: Boolean) = {
+      this(capacity=SequentialBlocks.DefaultCapacity, compress=compress)
+    }
+
+    /** The buffer storing the data. */
+    val byteBuffer = {
+      if (compress) ByteBuffer.allocate(MaxCompressedFixedBlockLength * capacity)
+      else ByteBuffer.allocate(DefaultUncompressedFixedBlockLength * capacity)
+    }
+    byteBuffer.order(ByteOrder.LITTLE_ENDIAN)
+
+    /** The number of blocks stored. */
+    var numStoredBlocks: Int = 0
+
+    /** The size of each stored block. */
+    var blockSizes = new ListBuffer[Int]() // NB: for uncompressed data, only the last size can be different
+
+    /** Gets the underlyng array of bytes. */
+    def array: Array[Byte] = this.byteBuffer.array()
+
+    /** Clears the underlying bytes, so this object can be re-used. */
+    def clear(): Unit = {
+      this.byteBuffer.clear()
+      blockSizes.clear()
+      numStoredBlocks = 0
+    }
+  }
+}
+
+/** Stores the input data to be processed, and the output data after processing.
+  *
+  * @param blockId the block identifier, typically the index of the block as it was read.
+  * @param compress true if the input data will be compressed, false otherwise.
+  */
+private class Blocks(var blockId: Long, compress: Boolean) extends Ordered[Blocks] {
+  import Blocks._
+
+  private val inData = new SequentialBlocks(compress = !compress)
+  private val outData = new SequentialBlocks(compress = compress)
+
+  // convenience methods for the input data
+  def inByteBuffer: ByteBuffer = inData.byteBuffer
+  def inBlockSizes: ListBuffer[Int] = inData.blockSizes
+  def inArray: Array[Byte] = inData.array
+  def incrementNumInBlocks(): Unit = inData.numStoredBlocks += 1
+
+  // convenience methods for the output data
+  def outByteBuffer: ByteBuffer = outData.byteBuffer
+  def outBlockSizes: ListBuffer[Int] = outData.blockSizes
+  def outArray: Array[Byte] = outData.array
+  def incrementNumOutBlocks(): Unit = outData.numStoredBlocks += 1
+
+  /** Clears the underling input and output data. */
+  def clear(): Unit = {
+    this.inData.clear()
+    this.outData.clear()
+  }
+
+  /** Compares two blocks by their id. */
+  def compare(that: Blocks): Int = this.blockId.compare(that.blockId)
+}
```

Comment:
Both this class and `SequentialBlocks` have a very struct-like feel to them, in that they seem to contain a bunch of state but don't have methods to mutate that state and instead (I'm assuming) expose their state for others to manipulate.  Makes it very hard to understand what's going on and whether things are maintained consistently (e.g. keeping the buffers, block sizes etc. in sync).